### PR TITLE
soc: arm: nordic_nrf: nrf53: Add Synchronized RTC

### LINKED ISSUE #23021: logging: Sync timestamp between cores
The logging system has a useful feature where a timestamp is prepended each log message.

This works well on single core MCU's, but in some situations you want to investigate multi-core applications. E.g. both cores crash, and you need to figure ut what took place first.
Suggest adding a config of type CONFIG_LOG_TIMESTAMP_SYNC. This will ensure that e.g. a LOG_DBG which is called on each core at the same point in time gets the same timestamp.

I realise that this is may well be far from trivial, as well as platform specific.
@nordic-krch

### PR DESCRIPTION
**Pull request by nordic-krch**
The goal of this PR is to establish a mechanism which allows to get the offset between system clock on network and application core (assuming they both run from RTC). Offset can be applied to a function which returns timestamp for logging allowing logs synchronization from both cores (currently on independent outputs).

Add a module which is responsible for getting offset between RTCs
used for system clock on NET and APP cores. After getting an offset
between NET and APP clocks, it can be used for logging timestamping
on NET core to ensure that timestamping is in sync on both cores.
    
Synchronization is done using PPI, IPM task and events and RTC
capture feature.

Contains also extension to `nrf_rtc_timer` by adding function for getting CAPTURE task address `z_nrf_rtc_timer_capture_task_address_get()` (not present on all nRF platforms).

Additionally added sample that shows results of RTC synchronization for nrf53 `samples/boards/nrf/nrf53_sync_rtc`

Fixes #23021

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

File: samples/boards/nrf/nrf53_sync_rtc/README.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+For simplicity and low latency sample is not using more sophisticated IPM protocols.
+
+Building the application for nrf5340dk_nrf5340_cpuapp
+*****************************************************
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/boards/nrf/nrf53_sync_rtc
+   :board: nrf5340dk_nrf5340_cpuapp
+   :goals: flash
+   :goals: flash --hex-file build/sync_rtc_net-prefix/src/sync_rtc_net-build/zephyr/zephyr.hex
```
Comment (nordic-krch) [inline samples/boards/nrf/nrf53_sync_rtc/README.rst:32]: @mbolivar not sure if there is a better way to flash `ExternalProject`. In ninja i can type `ninja sync_rtc_net flash` but couldn't find west equivalent.

**Comment by koffes** [comment]
Fantastic feature! I will have a look at the presentation as soon as I have time.

File: soc/arm/nordic_nrf/nrf53/sync_rtc.c [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+	nrfx_gppi_channels_enable(BIT(channels.ch.ppi));
+
+	if (IS_ENABLED(CONFIG_SOC_NRF5340_CPUAPP)) {
+		ppi_ipc_to_rtc(channels, true);
+	} else {
+		ppi_rtc_to_ipc(channels, true);
+		sync_cc = z_nrf_rtc_timer_read() + RTC_SYNC_ARBITRARY_OFFSET;
+		z_nrf_rtc_timer_compare_set(channels.ch.rtc, sync_cc,
+					    rtc_cb, (void *)channels.raw);
```
Comment (anangl) [inline soc/arm/nordic_nrf/nrf53/sync_rtc.c:233]: The use of `RTC_SYNC_ARBITRARY_OFFSET` here is a bit confusing, as the value used here has nothing to do with offset between APP and NET, it's only about successful setting of the RTC compare in the near future, right? And I think you should prevent preemption of this part of code.

File: include/drivers/timer/nrf_rtc_timer.h [hunk 3]
```diff
@@ -7,6 +7,8 @@
 #ifndef ZEPHYR_INCLUDE_DRIVERS_TIMER_NRF_RTC_TIMER_H
 #define ZEPHYR_INCLUDE_DRIVERS_TIMER_NRF_RTC_TIMER_H
 
+#include <sys_clock.h>
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/drivers/timer/nrf_rtc_timer.h:10]: it is somehow related. `z_nrf_rtc_timer_get_ticks` is using `k_timeout_t` from `sys_clock.h` so it was relying on it being included somewhere else. I can move it to separate commit.

File: include/drivers/timer/nrf_rtc_timer.h [hunk 4]
```diff
@@ -46,6 +48,18 @@ uint32_t z_nrf_rtc_timer_read(void);
  */
 uint32_t z_nrf_rtc_timer_compare_evt_address_get(int32_t chan);
 
+/** @brief Get CAPTURE task register address.
+ *
+ * Address can be used for (D)PPI.
+ *
+ * @note Not all platforms have CAPTURE task.
+ *
+ * @param chan Channel ID between 0 and CONFIG_NRF_RTC_TIMER_USER_CHAN_COUNT.
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/drivers/timer/nrf_rtc_timer.h]: added protection, updated doc

File: soc/arm/nordic_nrf/nrf53/sync_rtc.c [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+#include <drivers/timer/nrf_rtc_timer.h>
+#include <drivers/ipm.h>
+#include <logging/log_ctrl.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(sync_rtc, CONFIG_SYNC_RTC_LOG_LEVEL);
+
+#define RTC_SYNC_ARBITRARY_OFFSET 100
+
+static uint32_t sync_cc;
+uint32_t nrf53_sync_offset;
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply soc/arm/nordic_nrf/nrf53/sync_rtc.c]: added api `z_nrf_rtc_timer_nrf53net_rtc_offset` to `nrf_rtc_timer.h` but with implementation in `sync_rtc.c`

**Comment by nordic-krch** [comment]
@anangl comments applied.

**Comment by nordic-krch** [comment]
@anangl ping.

**Comment by nordic-krch** [comment]
@anangl ping

**Comment by nordic-krch** [comment]
Reworked to support use of mbox driver as well.

**Comment by nordic-krch** [comment]
@alexsven @thebatfisk @koffes can anyone of you try it if it works ok in your application. There is only simple example running and it would be good to validate it in more complex application.

**Review by anangl — APPROVED** [review]

File: soc/arm/nordic_nrf/nrf53/sync_rtc.c [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+	return mbox_set_enabled(&channel, true);
+}
+
+static int ipm_rx_init(void *user_data)
+{
+	const struct device *ipm_dev;
+
+	ipm_dev = device_get_binding("IPM_" STRINGIFY(CONFIG_NRF53_SYNC_RTC_IPM_IN));
+	if (ipm_dev == NULL) {
+		return -EIO;
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply soc/arm/nordic_nrf/nrf53/sync_rtc.c]: done

File: soc/arm/nordic_nrf/nrf53/sync_rtc.c [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+/* Setup or clear connection from RTC_COMPARE to IPC_SEND
+ *
+ * @param channels Details about channels
+ * @param setup If true connection is setup, else it is cleared.
+ */
+static void ppi_rtc_to_ipc(union rtc_sync_channels channels, bool setup)
+{
+
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply soc/arm/nordic_nrf/nrf53/sync_rtc.c]: done

File: soc/arm/nordic_nrf/nrf53/sync_rtc.c [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+ */
+
+/* Setup or clear connection from IPC_RECEIVE to RTC_CAPTURE
+ *
+ * @param channels Details about channels
+ * @param setup If true connection is setup, else it is cleared.
+ */
+static void ppi_ipc_to_rtc(union rtc_sync_channels channels, bool setup)
+{
+
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply soc/arm/nordic_nrf/nrf53/sync_rtc.c]: done

File: include/drivers/timer/nrf_rtc_timer.h [hunk 9]
```diff
@@ -154,6 +154,18 @@ void z_nrf_rtc_timer_abort(int32_t chan);
...
+ *
+ * Returned value added to the current system tick on network cpu gives current
+ * application cpu system tick. Function can only be used on network cpu. It
+ * requires @ref CONFIG_NRF53_SYNC_RTC being enabled.
+ *
+ * @retval Non-negative offset given in RTC ticks.
+ * @retval -ENOSYS if operation is not supported.
+ * @retval -EBUSY if synchronization is not yet completed.
+ */
+int z_nrf_rtc_timer_nrf53net_rtc_offset(void);
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/drivers/timer/nrf_rtc_timer.h]: done

**Review by carlescufi — APPROVED** [review]

**Pull request merged**